core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -188,16 +189,33 @@ private void cleanExpiredFiles(List<Snapshot> snapshots, Set<Long> validIds, Set
     // physically deleting files that were logically deleted in a commit that was rolled back.
     Set<Long> ancestorIds = Sets.newHashSet(SnapshotUtil.ancestorIds(base.currentSnapshot(), base::snapshot));
```

Comment:
@rdblue I can see that this PR handles almost all scenarios in the current/first iteration (when you are expiring some or any snapshot), but I am trying to understand the run following the current iteration, what is the output of this (SnapshotUtil.ancestorIds) API (on subsequent runs), after we expire one of the ancestorIds in the current run?

Based on what I can understand, it will return the ancestors based of off current snapshot, till the time we can find the parent snapshot in snapshot list, and since we have expired a particular ancestor (in previous run) the list will be incomplete. Right? If yes, then I suppose, we can't support APIs like expireSnapshotId, for selectively expiring snapshots, as it breaks lineage.
I gave a thought, and found out, that we can't support id based expiring, instead we can only do time-based expiry. That will elevate the issue, where in (See below example) I can expire C (cherrypicked from B), and then later decide to expire B, which might delete underlying dataFiles of B referred by C, which is expired in previous run. Let me know, how this PR solves that? 
//A -- C (B) -- D
//   `- B

Correct me if I am wrong, I am still trying to understand nuances of snapshot list maintenance in Iceberg.

Reply:
The current set of changes makes some assumptions that generally hold:
* Cherry-picked snapshots are newer than the original snapshot -- C comes after B in time
* Expiration happens in order -- B is expired before C or in the same commit

I'm not yet sure how to handle more complex cases, like expiring a specific snapshot. I think maybe we should just add a warning or deprecate that method. I don't know of any current uses.

Reply:
@rdblue +1, I am too in the opinion, that we should deprecate the older API of expiring a particular snapshotId, supporting it with cherry-pick (or more complex linking) is going to be tricky. 
Apologies I didn't followed on this issue, I was only stuck, since I also didn't had any solution for supporting expiry of a particular snapshot. I can help with the UTs though, I attempted to add some with my PR https://github.com/rominparekh/incubator-iceberg/pull/3/files